File: src/bokeh/models/misc/group_by.py
#-----------------------------------------------------------------------------
# Copyright (c) Anaconda, Inc., and Bokeh Contributors.
# All rights reserved.
#
# The full license is in the file LICENSE.txt, distributed with this software.
#-----------------------------------------------------------------------------
""" Models for describing grouping behavior of collections of models. """

#-----------------------------------------------------------------------------
# Boilerplate
#-----------------------------------------------------------------------------
from __future__ import annotations

import logging # isort:skip
log = logging.getLogger(__name__)

#-----------------------------------------------------------------------------
# Imports
#-----------------------------------------------------------------------------

# Bokeh imports
from ...core.has_props import abstract
from ...core.properties import Instance, List, Required
from ...model import Model

#-----------------------------------------------------------------------------
# Globals and constants
#-----------------------------------------------------------------------------

__all__ = (
    "GroupByModels",
    "GroupByName",
)

#-----------------------------------------------------------------------------
# General API
#-----------------------------------------------------------------------------

@abstract
class GroupBy(Model):
    """ Base class for grouping behaviors. """

    # explicit __init__ to support Init signatures
    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)

class GroupByModels(GroupBy):
    """ Group models by manually predefined groups. """

    # explicit __init__ to support Init signatures
    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)

    groups = Required(List(List(Instance(Model))), help="""
    Predefined groups of models.
    """)

class GroupByName(GroupBy):
    """ Group models by their names (``Model.name`` property). """

    # explicit __init__ to support Init signatures
    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)

# TODO GroupByCustomJS(GroupBy)

#-----------------------------------------------------------------------------
# Dev API
#-----------------------------------------------------------------------------

#-----------------------------------------------------------------------------
# Private API
#-----------------------------------------------------------------------------

#-----------------------------------------------------------------------------
# Code
#-----------------------------------------------------------------------------

File: bokehjs/src/lib/models/misc/group_by.ts
import {Model} from "../../model"
import type * as p from "core/properties"
import {List, Ref} from "core/kinds"

export namespace GroupBy {
  export type Attrs = p.AttrsOf<Props>
  export type Props = Model.Props
}

export interface GroupBy extends GroupBy.Attrs {}

export abstract class GroupBy extends Model {
  declare properties: GroupBy.Props

  constructor(attrs?: Partial<GroupBy.Attrs>) {
    super(attrs)
  }

  abstract query_groups(models: Iterable<Model>, pool: Iterable<Model>): Iterable<Model[]>
}

export namespace GroupByModels {
  export type Attrs = p.AttrsOf<Props>
  export type Props = GroupBy.Props & {
    groups: p.Property<Model[][]>
  }
}

export interface GroupByModels extends GroupByModels.Attrs {}

export class GroupByModels extends GroupBy {
  declare properties: GroupByModels.Props

  constructor(attrs?: Partial<GroupByModels.Attrs>) {
    super(attrs)
  }

  static {
    this.define<GroupByModels.Props>({
      groups: [ List(List(Ref(Model))) ],
    })
  }

  *query_groups(models: Iterable<Model>, _pool: Iterable<Model>): Iterable<Model[]> {
    for (const model of models) {
      for (const group of this.groups) {
        if (group.includes(model)) {
          yield group
        }
      }
    }
  }
}

export namespace GroupByName {
  export type Attrs = p.AttrsOf<Props>
  export type Props = GroupBy.Props
}

export interface GroupByName extends GroupByName.Attrs {}

export class GroupByName extends GroupBy {
  declare properties: GroupByName.Props

  constructor(attrs?: Partial<GroupByName.Attrs>) {
    super(attrs)
  }

  *query_groups(models: Model[], pool: Model[]): Iterable<Model[]> {
    const groups = new Map<string, Set<Model>>()
    for (const model of pool) {
      const {name} = model
      if (name != null) {
        let group = groups.get(name)
        if (group === undefined) {
          group = new Set()
          groups.set(name, group)
        }
        group.add(model)
      }
    }

    for (const model of models) {
      for (const group of groups.values()) {
        if (model.name != null && group.has(model)) {
          yield [...group]
        }
      }
    }
  }
}
